docs: update `universal` guide and docs example to v9

gkalpak opened the pull request:
This PR updates the `universal` guide and docs example to be aligned with what the v9 CLI generates. It also removes an obsolete section (and the corresponding code) and adds some tests for the example app.

See the individual commit messages for more details.

Fixes #35289.

##
This was originally part of #34374 (and split off for easier reviewing).
Partially addresses: [FW-1609](https://angular-team.atlassian.net/browse/FW-1609)

mary-poppins commented:
You can preview 0fccbbf at https://pr36483-0fccbbf.ngbuilds.io/.

mary-poppins commented:
You can preview 8516692 at https://pr36483-8516692.ngbuilds.io/.

IgorMinar commented on aio/content/guide/universal.md:
```diff
@@ -334,3 +266,37 @@ The following Node.js Express code routes all remaining requests to `/dist`, and
 
 <code-example path="universal/server.ts" header="server.ts (static files)" region="static"></code-example>
+
+### Using absolute URLs for server requests
```
```suggestion
### Using absolute URLs for HTTP (data) requests on the server
```

gkalpak replied:
Fixed.

IgorMinar commented on aio/content/examples/universal/src/app/absolute-url.interceptor.ts:
```diff
@@ -0,0 +1,34 @@
+      @Inject('SOME_TOKEN_UNDER_WHICH_THE_FULL_URL_IS_PROVIDED') @Optional()
+      private fullUrl: URL | null) {}
+
+  intercept(req: HttpRequest<any>, next: HttpHandler) {
```
```suggestion
  intercept(req: HttpRequest<unknown>, next: HttpHandler) {
```

Can you please use `unknown` whenever we can instead of `any` as `any` can lead to lots of type inference issues... thanks!

gkalpak replied:
I just move the code snippet from the guide to a file, so I kept the original code. I assume `any` was used to match the [HttpInterceptor#intercept()][1] signature (which features `any`).

In any case, this code is removed in the following fixup commit, so this doesn't really matter :grin:
I'll merge the fixup commit into this one to avoid confusion.

[1]: https://angular.io/api/common/http/HttpInterceptor#intercept

IgorMinar reviewed (approved):
overall LGTM! thanks!

IgorMinar reviewed (approved):
Reviewed-for: global-approvers

but let's have @aikidave also take a quick look to check the language. thanks!

gkalpak committed:
refactor(docs-infra): update `universal` example to match latest CLI

Update the `universal` example (and related files) to match what would
be generated by the latest CLI.

Fixes #35289

gkalpak committed:
build(docs-infra): update `@types/express-serve-static-core` to avoid error in `universal` example

Previously, building the `universal` example failed with:
```
node_modules/@types/express/index.d.ts(90,50): error TS2694: Namespace '".../@types/express-serve-static-core/index"' has no exported member 'Params'.
node_modules/@types/express/index.d.ts(90,64): error TS2694: Namespace '".../@types/express-serve-static-core/index"' has no exported member 'ParamsDictionary'.
```

This commit fixes the error by upgrading
`@types/express-serve-static-core` to a newer version.
See DefinitelyTyped/DefinitelyTyped#40905 for more details.

gkalpak committed:
test(docs-infra): add tests for `universal` docs example

Previously, there were no tests for the `universal` docs example. This
meant that the project was not tested at all (not even ensuring that it
can be built successfully).

This commit adds e2e tests for the `universal` example (ported from
`toh-pt6` and cleaned up) and also verifies that the project can be
built successfully (including the server).

gkalpak committed:
docs: minor fixes/improvements in the `universal` guide

gkalpak committed:
docs: remove unneeded code from `universal` example/guide

In the past, server-side rendered apps needed to convert URLs used in
API requests to absolute when rendering on the server. Originally, this
was handled in the `universal` guide and corresponding example app by
modifying the `HeroService` to use `APP_BASE_HREF` to derive the
absolute URL.

In #28956, the guide was updated to show an improved method: Specifying
an `HttpInterceptor` that took care of converting the URLs to absolute.
That interceptor was only provided when rendering the app on the server.
By mistake, the corresponding example app was not updated along with the
guide.

Since `@nguniversal/*` v7.1.0, it is no longer necessary to convert the
URLs to absolute inside the app. This is handled in the `@nguniversal`
libs (see angular/universal#897).

This commit updates the example app to remove unnecessary code and
modifies the guide to mention the issue with absolute URLs, but explain
that developers only need to worry about it when not using one of the
`@nguniversal/*-engine` packages.

gkalpak committed:
fixup! docs: remove unneeded code from `universal` example/guide

mary-poppins commented:
You can preview 11cbc21 at https://pr36483-11cbc21.ngbuilds.io/.

aikidave commented on aio/content/guide/universal.md:
```diff
@@ -328,13 +256,32 @@ The following code filters for request URLs with no extensions and treats them a
+This means that the URLs must be somehow converted to absolute when running on the server and be left relative when running in the browser.
+
+If you are using one of the `@nguniversal/*-engine` packages (such as `@nguniversal/express-engine`), this is taken care for you automatically.
+So, you don't need to do anything to make relative URLs work on the server.
```
```suggestion
You don't need to do anything to make relative URLs work on the server.
```

aikidave commented on aio/content/guide/universal.md:
```diff
@@ -328,13 +256,32 @@ The following code filters for request URLs with no extensions and treats them a
+If, for some reason, you are not using an `@nguniversal/*-engine` package, you may need to handle it yourself.
+
+There are several ways to achieve that.
+The recommended solution, which is also the least intrusive as it does not require any changes to the app, is to pass the full request URL to the `options` argument of [renderModule()](api/platform-server/renderModule) or [renderModuleFactory()](api/platform-server/renderModuleFactory) (depending on what you use to render `AppServerModule` on the server).
```
```suggestion
The recommended solution is to pass the full request URL to the `options` argument of [renderModule()](api/platform-server/renderModule) or [renderModuleFactory()](api/platform-server/renderModuleFactory) (depending on what you use to render `AppServerModule` on the server). This option is the least intrusive as it does not require any changes to the app, 
```

aikidave commented on aio/content/guide/universal.md:
```diff
@@ -328,13 +256,32 @@ The following code filters for request URLs with no extensions and treats them a
+Here, "request URL" refers to the URL of the request as a response to which the app is being rendered on the server.
+For example, if the client requested `https://my-server.com/dashboard` and you are rendering the app on the server to respond to that request, `options.url` should be set to `https://my-server.com/dashboard`.
+
+Now, on every HTTP request made as part of rendering the app on the server, Angular will be able to correctly resolve the request URL to an absolute URL, using the provided `options.url`.
```
```suggestion
Now, on every HTTP request made as part of rendering the app on the server, Angular can correctly resolve the request URL to an absolute URL, using the provided `options.url`.
```

aikidave commented on aio/content/guide/universal.md:
```diff
@@ -328,13 +256,32 @@ The following code filters for request URLs with no extensions and treats them a
+
+If, for some reason, you are not using an `@nguniversal/*-engine` package, you may need to handle it yourself.
+
+There are several ways to achieve that.
```
Unless I am missing something, it appears that we say that there are several ways to handle relative URLs, but only provide one, as it is the recommended way to do it. If that's the case, we can delete line 282.

aikithoughts reviewed (commented):
A few suggestions to make the text a little shorter and easier to read.

gkalpak committed:
build(docs-infra): add missing `build` npm script for `universal` docs example ZIP archive

Previously, the `package.json` template used in the ZIP archive of the
`universal` example that we offer for download missed the `build` npm
script.

This commit updates the template for the `universal` docs example to
include the `build` npm script.

NOTE:
The `build` npm script is already included in
`aio/tools/examples/shared/boilerplate/universal/package.json`, but it
was removed by the example zipper.

gkalpak committed:
style(docs-infra): clean up Tour-of-Heroes examples

I noticed these minor styling issues while aligning the `universal`
examples with the `toh-pt6` example (in a subsequent commit).

gkalpak committed:
refactor(docs-infra): update `main.ts` in Tour-of-Heroes examples to match latest CLI

Update the `main.ts` files in Tour-of-Heroes examples to match what
would be generated by the latest CLI.

gkalpak committed:
fix(docs-infra): align `universal` example with `toh-pt6`

As mentioned in the `universal` guide, the `toh-pt6` examples is the
starting poitn for the `universal` example. However, the two examples
had become out-of-sync, because some fixes/changes were made to the
Tour-of-Heroes examples.

This commit ports these changes to the `universal` example.

gkalpak committed:
fixup! docs: remove unneeded code from `universal` example/guide

gkalpak commented:
Thx, @aikidave! I've made the changes you suggested.

mary-poppins commented:
You can preview 928dd02 at https://pr36483-928dd02.ngbuilds.io/.

alan-agius4 reviewed (approved):
LGTM, from universal side. Thanks @gkalpak.
